src/tests/suite/daos_dist_tx.c

```diff
@@ -0,0 +1,986 @@
+	struct ioreq	 req;
+
+	print_message("Read only transaction\n");
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_SINGLE, arg);
+
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, DAOS_TX_NONE,
+		      &req);
+
+	insert_single(dkey2, akey, 0, write_buf, DTX_IO_SMALL, DAOS_TX_NONE,
+		      &req);
+
+	MUST(daos_tx_open(arg->coh, &th, DAOS_TF_RDONLY, NULL));
+
+	req.arg->expect_result = -DER_NO_PERM;
+
+	insert_single(dkey, akey2, 0, write_buf, DTX_IO_SMALL, th, &req);
+	punch_akey(dkey, akey, th, &req);
+
+	req.arg->expect_result = 0;
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+	assert_int_equal(req.iod[0].iod_size, DTX_IO_SMALL);
+
+	lookup_single(dkey, akey2, 0, fetch_buf, DTX_IO_SMALL, th, &req);
+	assert_int_equal(req.iod[0].iod_size, 0);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_12(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	const char	*akey2 = "tmp_akey";
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+	int		 i;
+
+	print_message("DTX with zero copy flag\n");
+	MUST(daos_tx_open(arg->coh, &th, DAOS_TF_ZERO_COPY, NULL));
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+	insert_single(dkey, akey, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	/* Reuse 'write_buf' */
+	for (i = 0; i < DTX_IO_SMALL; i++)
+		write_buf[i] += 1;
+
+	insert_single(dkey, akey2, 0, write_buf, DTX_IO_SMALL, th, &req);
+
+	MUST(daos_tx_commit(th, NULL));
+
+	lookup_single(dkey, akey2, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	assert_memory_equal(write_buf, fetch_buf, DTX_IO_SMALL);
+
+	lookup_single(dkey, akey, 0, fetch_buf, DTX_IO_SMALL,
+		      DAOS_TX_NONE, &req);
+	/* 'write_buf' has been overwritten, so it is the same as 2nd update. */
+	assert_memory_equal(write_buf, fetch_buf, DTX_IO_SMALL);
+
+	ioreq_fini(&req);
+	MUST(daos_tx_close(th, NULL));
+}
+
+static void
+dtx_13(void **state)
+{
+	test_arg_t	*arg = *state;
+	const char	*dkey = dts_dtx_dkey;
+	const char	*akey = dts_dtx_akey;
+	char		 write_buf[DTX_IO_SMALL];
+	char		 fetch_buf[DTX_IO_SMALL];
+	daos_handle_t	 th = { 0 };
+	daos_obj_id_t	 oid;
+	struct ioreq	 req;
+	int		 rc;
+
+	print_message("DTX status machnie\n");
+
+	arg->async = 0;
+	oid = dts_oid_gen(OC_RP_XSF, 0, arg->myrank);
+	ioreq_init(&req, arg->coh, oid, DAOS_IOD_ARRAY, arg);
+	dts_buf_render(write_buf, DTX_IO_SMALL);
+
+	print_message("Open the TX1...\n");
+	MUST(daos_tx_open(arg->coh, &th, 0, NULL));
+
+	print_message("Commit the empty TX1...\n");
+	MUST(daos_tx_commit(th, NULL));
+
+	print_message("Commit the committed TX1...\n");
```

Comment:
I am not against to return -DER_ALREADY for such case, but I am not sure about our TX commit semantics: whether allow to repeatedly commit the same TX or not.

Currently, on server side, try to commit a committed DTX entry will NOT generate error because there is race between batched committed ULT and other regular ULT with commit-on-share mechanism.

Then should we follow consistent logic for client side TX commit?

Reply:
I don't have a strong opinion here either, since this appears harmless from the user perspective since updates will not be allowed (fetch - in discussion).
I just lean to return an error since my understanding is that a successful commit or abort should render the tx handle unusable anymore other than freeing the handle with close.

Reply:
>  a successful commit or abort should render the tx handle unusable anymore other than freeing the handle with close.

I am not sure for that. If that is true, then why we need separated commit/abort/close API? only one "close(tx, result)" API is enough. My understand is that, we should allow other operations after the TX committed, that is the benefit of offering separated "commit()" API.

Reply:
yes i originally designed that close to be separate from commit to allow sharing a TX handle between multiple ranks with g2l, l2g like we do for containers, pool.
this is not supported today though.
but the idea is 1 process commits  then closes, then other can just call close.
otherwise we don't really need tx_close() api

Reply:
tx_hdl2epoch() also touch the TX handle after commit.

Reply:
that is fine really, because you are not doing anything "with" the tx, but more like getting some info that corresponds to the tx.

Reply:
Fine, let's make more discussion about whether allow read against commit TX handle in other channel.